packages/filesystem/src/browser/file-tree/file-tree-widget.ts

```diff
@@ -56,4 +66,140 @@ export class FileTreeWidget extends TreeWidget {
+                }
+            }
+        }
+    }
+
+    protected handleDirectory(webDirEntry: WebKitEntry, dest: string) {
+
+        const getEntries = (dirEntry: any, destination: string) => {
+            const reader = dirEntry.createReader();
+            reader.readEntries((e: any) => {
+                const results = <WebKitEntry[]>e;
+
+                for (let count = 0; count < results.length; count++) {
+
+                    if (results[count].isDirectory) {
+                        getEntries(results[count], destination);
+                    } else {
+                        const fileEntry = <WebKitFileEntry>results[count];
+                        fileEntry.file(e => {
+                            const f = <File>(e as {});
+                            this.buildAndSendData(destination + dirEntry.fullPath, f, f.name);
+                        });
+                    }
+                }
+            }, (error: any) => {
+                this.logger.error(error.toString());
+            });
+        };
+        getEntries(webDirEntry, dest);
+    }
+
+    protected buildAndSendData(destination: string, file: File, fileName: string) {
+        const fileToCreate = destination + "/" + fileName;
+        this.fileSystem.exists(fileToCreate).then(exist => {
+            this.fileSystem.createFile(fileToCreate).then(() => {
+
+                // Saving contents
+                this.fileSystem.getFileStat(fileToCreate).then(currentStat => {
+                    const reader = new FileReader();
+
+                    reader.onload = e => {
+                        const fileContent: ArrayBuffer = reader.result;
+                        const buffer64 = base64.fromByteArray(new Uint8Array(fileContent));
+                        this.fileSystem.setContentBase64(currentStat, buffer64).then(() => Promise.resolve());
+                    };
+                    reader.readAsArrayBuffer(file);
+                });
+            },
+                (error: Error) => {
+                    // Error: Error occurred while creating the file. File already exists
+                    // Ask the end-user to replace the file or not
+                    this.shouldOverwrite(fileToCreate).then(replace => {
+                        if (replace) {
+                            // Rename the existing file with a new extension.
+                            const mvFile = fileToCreate + ".orig";
+                            const overwrite = { overwrite: true };
+                            this.fileSystem.move(fileToCreate, mvFile, overwrite);
+
+                            // Bring the new file into the folder
+                            this.buildAndSendData(destination, file, fileName);
+                        }
+                    });
+                });
+        });
+    }
+
+    // Create a dialog to ask the user intention about overwriting the file
+    protected async shouldOverwrite(file: string): Promise<boolean> {
```

Comment:
it should be done only once by `FileSystemLister`, the design should not require duplications

Reply:
Could we skip backing up of existing files for now and just override it using the generic logic, or it is very important to preserve old files?